### DISCUSSION TITLE
Tectonic setup in VSCode with Latex Workshop

### DISCUSSION THREAD

**Discussion by thuvasooriya** [tectonic-typesetting/tectonic#1181] [Q&A]
I've recently found tectonic and fell in love with it as it compiles my somewhat complex latex project with just one command with no effort. CLI and the tool is topnotch but I had to go through what I would call a trial by fire to set it up in vscode with latex workshop to make it work somewhat reliable. Noob alert. It took me about 2 days. Most of it has to do with me and the dumb decisions I made.

I couldn't get tectonic command to get detected by the extension due to some path issue with the extension with nix installed packages which works in CLI anyway.

I have referenced the other discussion here about the VSCode setup and a whole lot of reading docs and this is what I could come up with.

```
 {
...
    "latex-workshop.latex.recipe.default": "tectonic",
    "latex-workshop.latex.autoBuild.run": "onSave",
    "latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/%DOCFILE%",
    "latex-workshop.view.pdf.viewer": "tab",
    "latex-workshop.latex.recipes": [
        {
            "name": "tectonic",
            "tools": [
                "tectonic"
            ]
        }
    ],
    "latex-workshop.latex.tools": [
        {
            "name": "tectonic",
            "command": "tectonic",
            "args": [
                "-X",
                "build",
                "--keep-intermediates",
                "--keep-logs"
            ],
            "env": {
                "PATH": "/etc/profiles/per-user/tony/bin",
                // "TEXMFHOME": "~/Library/TinyTeX/bin"
            }
        }
    ],
    "latex-workshop.latexindent.path": "/opt/homebrew/bin/latexindent",
}
```

I wanted to use TinyTeX if I must install another TeXLive since I didn't know any better to fix the latexindent error but finally had to install a homebrew package and set the path to make it work.
This is clearly not ideal and I'm somewhat stressed with the things I might be doing wrong.
[https://github.com/thuvasooriya/tectomin](url)

Please point out the things I can do to improve this to flush the configuration to make it work with minimal effort in a new machine.
Some specific issues I faced: 
What should I do to make other latex workshop features (like synctex and other tools) work?
Even when I change only a single chapter the whole pdf is compiling? i guess. I there anything I can do to make it better? or is it supposed to happen.

**Comment by Ectalite** [comment]
If you install Tectonic with [homebrew](https://formulae.brew.sh/formula/tectonic) on macOS, it will be automatically added to the PATH. You don't need to add the PATH to TinyTeX.

> "env": {
                "PATH": "/etc/profiles/per-user/tony/bin",
                // "TEXMFHOME": "~/Library/TinyTeX/bin"
            }

Actually, I think what you are doing here is breaking the PATH variable, just remove the env variables.

**Reply by thuvasooriya** [reply]
hi thanks for the reply.
yes path might be an issue but since i'm moving all my configurations to nix choosing homebrew to install something that is already available in nix is kind of redundant. but i get the point.

is there any specific formatting options that are provided by tectonic by default right now? rather than using latex indent?

**Reply by Ectalite** [reply]
With my very limited knowledge with Nix is would add these packages to the **/etc/nixos/configuration.nix** file:
```
environment.systemPackages = [
    pkgs.tectonic
    pkgs.texlivePackages.latexindent
];
```
And for vscode I would use this config file:
```
"latex-workshop.latex.recipes": [
        {
            "name": "tectonic",
            "tools": [
                "tectonic"
            ]
        }
    ],
    "latex-workshop.latex.tools": [
        {
            "name": "tectonic",
            "command": "tectonic",
            "args": [
                "-X",
                "build",
                "--keep-intermediates",
                "--keep-logs",
                "--synctex"
            ]
        }
    ],
```
So please remove the **latex-workshop.latexindent.path** setting and `env` parameter for your tectonic setting inside **latex-workshop.latex.tools**.

**Comment by Fellow-baka** [comment]
Hello!
I'm new to vscode+latex workshop and can't run it. It is fine when I run tectonic from command line/powershell.

The contents of my config is the same as above:

```
{
  "latex-workshop.latex.recipe.default": "tectonic",
  "latex-workshop.latex.autoBuild.run": "onSave",
  "latex-workshop.latex.outDir": "%WORKSPACE_FOLDER%/build/index",
  "latex-workshop.view.pdf.viewer": "tab",
  "latex-workshop.latex.recipes": [
    {
      "name": "tectonic",
      "tools": ["tectonic"]
    }
  ],
  "latex-workshop.latex.tools": [
    {
      "name": "tectonic",
      "command": "tectonic",
      "args": ["-X", "build", "--keep-intermediates", "--keep-logs"]
    }
  ]
}
```

The output is following:
```
[13:55:10.959][Commander] RECIPES command invoked.
[13:55:10.959][Commander] BUILD command invoked.
[13:55:10.959][Build] The document of the active editor: output:extension-output-James-Yu.latex-workshop-%231-LaTeX Workshop
[13:55:10.959][Build] The languageId of the document: latex_workshop_log
[13:55:10.959][Build] Cannot find LaTeX root file. See https://github.com/James-Yu/LaTeX-Workshop/wiki/Compile#the-root-file
```

Tectonic is in win path variable, I can invoke it vscode terminal with no problems and build the project from it.
The project I'm trying to build is the default "hello world one" generated with ``` tectonic -X new```

**Reply by Fellow-baka** [reply]
nvm, i'm just stupid, had to use it in index.tex
All works now

**Comment by HealthyPear** [comment]
See #1327 - works like a charm!
